clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -754,6 +765,24 @@ public void onSuccess(ClientResponse response, RequestFuture<ListOffsetResult> f
                 });
     }
 
+    private RequestFuture<OffsetsForLeaderEpochResponse> sendOffsetsForLeaderEpochRequest(
+            final Node node,
+            final Map<TopicPartition, OffsetsForLeaderEpochRequest.PartitionData> partitionMap) {
```

Comment:
@mumrah @hachikuji, It looks like this method is not used anywhere. If that is true, how is the client discovering the "largest epoch smaller than the current epoch"?

Reply:
This is called from `Fetcher#prepareFetchRequests`. 

How it works is really the crux of this PR 😄. It checks if the leader has changed (up around L931) by comparing the leader+epoch from latest validated fetch position (stored in the subscription state) with the current leader+epoch according to Metadata. Metadata gets update asynchronously and is what we use as a reference for the cluster state. 

So, if the leader changes, Metadata gets updated and the next time we fetch we will see that the epoch has changed and call `sendOffsetsForLeaderEpochRequest`. This sends an OffsetForLeader request (asynchronously) to the broker and then returns. The partitions transition into a "waiting for validation state" and cannot be fetched from until they are validated.

There is some request batching going on as well to make it a bit more efficient, but slightly confuses things.

In `sendOffsetForLeaderRequest` there is a big callback (`RequestFutureAdaptor#onSuccess`) which does the validation check and either clears the "waiting for validation" state or it truncates (or it throws an error if it can't truncate due to missing reset policy).

Reply:
I checked out this branch and grepped against it:
```
$ rg sendOffsetsForLeaderEpochRequest
clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
768:    private RequestFuture<OffsetsForLeaderEpochResponse> sendOffsetsForLeaderEpochRequest(
```

Reply:
Ah yea, sorry my mistake. I read that as `sendOffsetForLeaderRequest` which does exist. I'll fix